### PR TITLE
Enhancement for MNIST test (Travis error #568)

### PR DESCRIPTION
**Pull request by pliptor**
This changes the activation function of the last layer in the mnist example from tan_h to softmax.
The output of test results become more intuitive this way. Result accuracy verified in simulations.
Testing with bitmap sample verified as well. Complete test results attached.

accuracy:98.69% (9869/10000)
* 0 1 2 3 4 5 6 7 8 9
0 973 0 0 0 0 3 4 0 5 0
1 0 1129 3 0 0 0 2 2 0 2
2 0 0 1021 2 2 0 1 7 1 1
3 0 0 0 998 0 5 1 1 2 2
4 0 0 1 0 973 0 3 0 2 11
5 1 1 0 4 0 882 4 0 1 4
6 2 2 1 0 2 2 941 0 0 1
7 1 0 1 4 1 0 0 1015 2 8
8 2 3 5 2 0 0 2 0 960 3
9 1 0 0 0 4 0 0 3 1 977
./example_minist_test 4.bmp
4,99.4535
7,0.23164
2,0.173432

[mnist_train_upstream.txt](https://github.com/tiny-dnn/tiny-dnn/files/769026/mnist_train_upstream.txt)
#568

### CODE REVIEW

**Comment by beru** [comment]
@pliptor You need to use `clang-format` command. Please check https://github.com/tiny-dnn/tiny-dnn/blob/master/CONTRIBUTING.md

You don't need to create another PR in case travis-ci jobs fails. You may fix problems, commit, push.
People will assist you in the process.

**Comment by pliptor** [comment]
@beru Sorry, I'm new. I thought I messed up. I don't know if this is going to fail again but hopefully not.

**Comment by beru** [comment]
I'm not sure using softmax makes example_mnist_train program converge faster or give more accuracy since I haven't tested it myself.

Jut by looking the diff, LGTM

**Comment by pliptor** [comment]
@Randl Thanks! I'll try to help with the Wiki pages on the example too.

**Comment by pliptor** [comment]
@beru I believe convergence speed does not change much. At least looking at the progression of through the epochs. Similar comment for accuracy. I think it's mostly presentation of the output that's changing to a normalized form. I did some different tests, however, that appears do improve both convergence and accuracy.

**Comment by beru** [comment]
@pliptor It'd be helpful if you could kindly update `examples/mnist/readme.md` as well but not mandatory.

I think this should be out of scope of this PR but `examples/mnist/test.cpp` loads network model from a serialized file, so **ideally** there shouldn't be needed to explicitly specify layer type, activation function type to process test result. This is a matter of API.

**Comment by pliptor** [comment]
@beru Yes, I'd be glad to adjust examples/mnist/readme.md accordingly. 

You are right about that there should not be a need to specify the activation type in examples/mnist/test.cpp At the end of #566 I have some comments about this. One could in principle just write a normalization function irrespective to the activation function that replaces the scaling function in examples/mnist/test.cpp. I could argue that it would be probably most generic and proper. I may give it a try.

**Comment by beru** [comment]
@pliptor I tested `example_mnist_train` in your PR branch but I couldn't get good result with it so I need to cancel my LGTM.

Tested env:
  Intel(R) Core(TM) i7-6700HQ CPU @ 2.60GHz
  Ubuntu 16.04.1 LTS
  gcc 5.4.0

  $ cmake -DBUILD_EXAMPLES=ON -DUSE_TBB=ON

Tested revision:
  b18e8c4483b99d883853d9d4b30ad87cf339df65

[log.txt](https://github.com/tiny-dnn/tiny-dnn/files/769149/log.txt)

Can you please report the output of `example_mnist_train` as well?

**Comment by pliptor** [comment]
@beru Thanks for your time trying to cross check it. This is very strange. Please give me a moment to check the output of example_mnist_train.

**Comment by beru** [comment]
Let me also attach the output of the program with current master HEAD. (1cf64c8c1e2522dced701023a29da52c00b75fe9)

[log.txt](https://github.com/tiny-dnn/tiny-dnn/files/769152/log.txt)

Strangely, it doesn't reach 99% accuracy after 30 epochs.

**Comment by pliptor** [comment]
Mine is attached. It takes a good 50min to run it without AVX available but I can run again later to confirm.

[mnist_train_upstream.txt](https://github.com/tiny-dnn/tiny-dnn/files/769156/mnist_train_upstream.txt)

**Comment by pliptor** [comment]
Your results with HEAD is more in line with what I'm getting with this commit.

**Comment by beru** [comment]
@pliptor I see, thank you for the report. So there seems to be a problem with AVX code.

By the way, this output is slightly different setting with master HEAD. (1cf64c8c1e2522dced701023a29da52c00b75fe9)

```cpp
// in train.cpp
int minibatch_size = 16;

// in config.h
#define CNN_TASK_SIZE 32

```

[log.txt](https://github.com/tiny-dnn/tiny-dnn/files/769159/log.txt)

**Comment by pliptor** [comment]
Cool so we uncovered a bug with this exercise. I hope my commit doesn't have a problem. It does explain why the result on Wiki (not the scaling part) look bad as well. I think it was generated with AVX on.

**Comment by beru** [comment]
Tested PR branch `example_mnist_train` without AVX.
$ cmake -DBUILD_EXAMPLES=ON -DUSE_TBB=ON -DUSE_AVX=OFF
Tested revision: b18e8c4483b99d883853d9d4b30ad87cf339df65
[log.txt](https://github.com/tiny-dnn/tiny-dnn/files/769166/log.txt)

So, LGTM

**Comment by pliptor** [comment]
@beru Thanks. Our results match right on the dot now. Should I create a different pull for the  examples/mnist/readme.md changes?

**Comment by beru** [comment]
@pliptor 

> BTW I noticed yesterday that there are two separate sections on Wiki reporting results for MNIST. One on the left and one on the right. The big difference in predicted output for the same input test shows the problem was already present at that time.

https://github.com/tiny-dnn/tiny-dnn/tree/master/examples/mnist
https://github.com/tiny-dnn/tiny-dnn/wiki/MNIST-Example

I updated `examples/mnist/readme.md` before. At that time I thought it strange but hadn't paid further attention much.

Redundant Wiki page saves us.

> Should I create a different pull for the examples/mnist/readme.md changes?

It's up to you.

**Comment by pliptor** [comment]
@beru my results with

// in train.cpp
int minibatch_size = 16;

// in config.h
#define CNN_TASK_SIZE 32

[beru_opt.txt](https://github.com/tiny-dnn/tiny-dnn/files/769189/beru_opt.txt)
It runs faster and it is more accurate (98.87%) than this proposed commit (98.69%). It's not as good as your 99.1% so I guess tan_h gives a bit more accurate result than softmax in the last layer.

**Comment by edgarriba** [comment]
original Lecun's paper¹ uses `tan_h` activations

¹ http://yann.lecun.com/exdb/publis/pdf/lecun-98.pdf

**Comment by pliptor** [comment]
@edgarriba  Thanks for the reference! I'm learning a lot now and running additional tests.

**Comment by pliptor** [comment]
Additional tests showing so far that accuracy of test phase is better with tan_h than softmax in the last layer. I think I'll withdraw this merge request. I might do just a fix on the test code side as the scaling function is not doing what the comment in the code is though. Thanks everyone for looking at this.
